README.md
condor_restd
============
REST-based API for HTCondor, using the Python bindings.

Currently allows read-only queries for jobs (in-queue and historical),
configuration, and machine status.

NOTE: The API has not stabilized yet and queries are subject to change.
Do not use in production.


KBase Specific Configuration Information
-----------
* To specify the auth url set `AUTH_URL`
* To specify the jwt token used to connect to condor specify `CONDOR_JWT_TOKEN`
* To specify which condor collector, schedd, and which variables are redacted (e.g. `environment`), modify the condor_config either here, or mount it in via a ConfigMap
* This api is protected via AUTH Roles, and checks to see if you have the proper EE2 Role to allow you access to this api
* For information about where python and its dependencies are installed, see the `Dockerfile` and `entrypoint.sh`

Installation
------------
Installation requires the following Python modules:

- `htcondor >= 8.9.2`
- `flask`
- `flask-restful`

Install them via `pip` or your OS's package manager.

Create a virtualenv then `pip install -e .`.  To run using the
built-in Flask server (not for production), run

    FLASK_APP=condor_restd flask run -p 9680

For additional scalability, run using a WSGI server such as gunicorn:

    gunicorn -w4 -b127.0.0.1:9680 condor_restd:app


These commands will run the server on port 9680.

NOTE: Running the htcondor-restd as root is not recommended.  This
readme assumes that you are running the restd as the unprivileged user
`restd`.

Configuration
-------------
The RESTD is configured using HTCondor configuration; see
[the configuration section of the HTCondor manual](https://htcondor.readthedocs.io/en/latest/admin-manual/introduction-to-configuration.html)
for general configuration information.  Note that RESTD-specific
parameters are not listed in the manual yet.

If you do not have permission to edit global HTCondor configuration,
you may configure the RESTD in one of two ways, as described in the
[Ordered Evaluation to Set the Configuration](https://htcondor.readthedocs.io/en/latest/admin-manual/introduction-to-configuration.html#ordered-evaluation-to-set-the-configuration)
section of the HTCondor manual:

1. Adding the options to the user configuration file for the user
running the restd, in `$HOME/.config/user_config` (or the location
defined by `USER_CONFIG_FILE`).  This is recommended for non-container
installs.

2. Using the `_CONDOR_*` environment variables. For example, set
the value for `RESTD_MAX_JOBS` by setting the environment variable
`_CONDOR_RESTD_MAX_JOBS`.  This is recommended for container installs.

The following HTCondor configuration values apply:

- `CONDOR_HOST`: The hostname of the HTCondor central manager.
- `RESTD_HIDE_JOB_ATTRS`: A comma or space-separated list of job
  attributes to redact from queries.  Attributes in this list will
  show the value `<REDACTED>` in the jobs and history endpoints.
- `RESTD_MAX_JOBS`: The maximum number of jobs returned for jobs,
  grouped_jobs, history, and grouped_history queries.


Queries
-------
The following queries are implemented.  Arguments in brackets `{}` are optional:


### jobs and history

Access job information (similar to `condor_q` and `condor_history`).
`jobs` and `history` behave exactly the same, except `jobs` queries jobs in the queue,
and `history` queries jobs that have left the queue.

    GET /v1/jobs/{schedd}{/clusterid}{?projection,constraint}
    GET /v1/history/{schedd}{/clusterid}{?projection,constraint}

Returns a list of job objects.  A job object looks like

    {
      "jobid": "123.45",
      "classad": { <classad> }
    }

Returns an empty list if no jobs match.

`schedd` is the name of the schedd to query, or `DEFAULT` to use
the default schedd (if there is one). Raises `404` if the schedd
does not exist.

`clusterid` limits the results to jobs with the given cluster ID.

`projection` is one or more comma-separated attributes; if specified,
only those attributes will be in the `classad` object of each job.

`constraint` is a classad expression restricting which jobs to include
in the result.

    GET /v1/jobs/{schedd}/{clusterid}/{procid}{?projection}
    GET /v1/history/{schedd}/{clusterid}/{procid}{?projection}

Returns a single job object with cluster ID given by `clusterid` and
the proc ID given by `procid`.
Raises `404` if no such job exists.

`schedd` is the name of the schedd to query, or `DEFAULT` to use
the default schedd (if there is one). Raises `404` if the schedd
does not exist.

`projection` is one or more comma-separated attributes; if specified,
only those attributes will be in the `classad` object of the job.

    GET /v1/jobs/{schedd}/{clusterid}/{procid}/{attribute}
    GET /v1/history/{schedd}/{clusterid}/{procid}/{attribute}

Returns a single attribute of a job with cluster ID given by `clusterid`,
proc ID given by `procid`, and attribute name given by `attribute`.

`schedd` is the name of the schedd to query, or `DEFAULT` to use
the default schedd (if there is one). Raises `404` if the schedd
does not exist.

Raises `404` if no such job exists, or if the attribute is undefined.


### grouped_jobs and grouped_history

Like `jobs` and `history`, accesses job information.  However, they
group the returned jobs by an attribute.

`grouped_jobs` and `grouped_history` behave exactly the same, except
`grouped_jobs` queries jobs in the queue, and `grouped_history`
queries jobs that have left the queue.

    GET /v1/grouped_jobs/{schedd}/{groupby}{/clusterid}{?projection,constraint}
    GET /v1/grouped_history/{schedd}/{groupby}{/clusterid}{?projection,constraint}

Returns an object of lists of job objects, keyed by the value of the
attribute given in `groupby`.  A job object looks like:

    {
      "jobid": "123.45",
      "classad": { <classad> }
    }

The returned object looks like:

    {
      "value1": [ <job objects> ],
      "value2": [ <job objects> ]
    }

Returns an empty object if no jobs match.  Jobs that do not have the
attribute given in `groupby` are omitted from the result.  (This is
because null is not a valid key.)

`schedd` is the name of the schedd to query, or `DEFAULT` to use
the default schedd (if there is one). Raises `404` if the schedd
does not exist.

`clusterid` limits the results to jobs with the given cluster ID.

`projection` is one or more comma-separated attributes; if specified,
only those attributes, plus the `groupby` attribute, will be in the
`classad` object of each job.

`constraint` is a classad expression restricting which jobs to include
in the result.


### config

Access config information (similar to `condor_config_val`).

    GET /v1/config{/attribute}{?daemon}

If `attribute` is specified, returns the value of the specific
attribute in the condor config.  If not specified, returns an object
of the form:

    {
      "attribute1": "value1",
      "attribute2": "value2",
      ...
    }

If `daemon` is specified, query the given running daemon; otherwise,
query the static config files.

Returns 404 if `attribute` is specified but the attribute is undefined.


### status

Access machine and daemon information (similar to `condor_status`).

    GET /v1/status{/name}{?projection,constraint,query}

This returns an array of objects of the following form:

    {
      "name": "<name classad attribute>",
      "type": "<ad type>",
      "classad": { <classad object> }
    }

`name` is a specific host or slot to query.  If not specified, all
matching ads are returned.

`query` is the type of ad to query; see the "Query options" in the
condor_status(1) manpage.  "any" is the default.

`projection` is one or more comma-separated attributes; if specified,
only those attributes will be in the `classad` object of each job.

`constraint` is a classad expression restricting which ads to include
in the result.


### grouped_status

Like `status`, accesses machine and daemon information.  However, it
groups the returned objects by a classad attribute.

    GET /v1/grouped_status/{groupby}{/name}{?projection,constraint,query}

Returns an object of lists of status objects, keyed by the value of
the attribute given in `groupby`.  A status object looks like:

    {
      "name": "<name classad attribute>",
      "type": "<ad type>",
      "classad": { <classad object> }
    }

The returned object looks like:

    {
      "value1": [ <status objects> ],
      "value2": [ <status objects> ]
    }

Returns an empty object if nothing matches.  Status objects that do
not have the classad attribute given in `groupby` are omitted from
the result.  (This is because null is not a valid key.)

`name` is a specific host or slot to query.  If not specified, all
matching ads are returned.

`query` is the type of ad to query; see the "Query options" in the
condor_status(1) manpage.  "any" is the default.

`projection` is one or more comma-separated attributes; if specified,
only those attributes, plus the `groupby` attribute, will be in the
`classad` object of each job.

`constraint` is a classad expression restricting which ads to include
in the result.


RELEASE_NOTES.md
# Release Notes
# 1.0.0
* Had to pin python versions due to this issue https://stackoverflow.com/questions/72157708/flask-restx-request-parser-returns-400-bad-request/72174259#72174259
* Adds authentication via auth roles
* Adds support to add JWT token via ENV variables



condor_restd.py
# -*- coding=utf-8 -*-
"""*** condor_restd.py ***
REST-based API for HTCondor, based on the HTCondor Python bindings.

Allows read-only queries for jobs (in-queue and historical),
configuration, and machine status.
"""
import re
import json

try:
    from typing import Dict, List
except ImportError:
    pass

import classad
import htcondor
from htcondor import AdTypes, Collector, DaemonTypes, RemoteParam

from flask import Flask
from flask_restful import Resource, Api, abort, reqparse

import utils

app = Flask(__name__)
api = Api(app)


def validate_attribute(attribute):
    """Return True if the given attribute is a valid classad attribute name"""
    return bool(re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", attribute))


def validate_projection(projection):
    """Return True if the given projection has a valid format, i.e.
    is a comma-separated list of valid attribute names.
    """
    return all(validate_attribute(x) for x in projection.split(","))


class JobsBaseResource(Resource):
    """Base class for endpoints for accessing current and historical job
    information. This class must be overridden to specify `executable`.

    """

    querytype = None

    def query(self, clusterid, procid, constraint, projection, attribute):
        schedd = utils.get_schedd()
        requirements = "true"
        if clusterid:
            requirements += " && clusterid==%d" % clusterid
        if procid:
            requirements += " && procid==%d" % procid
        if constraint:
            requirements += " && %s" % constraint

        if attribute:
            if not validate_attribute(attribute):
                abort(400, message="Invalid attribute")
            projection = [attribute, "clusterid", "procid"]
        elif projection:
            if not validate_projection(projection):
                abort(
                    400,
                    message="Invalid projection: must be a comma-separated list of classad attributes",
                )
            projection = set(projection.split(","))
            projection.add("clusterid")
            projection.add("procid")
        else:
            projection = set()

        if self.querytype == "history":
            classads = schedd.history(
                requirements=requirements, projection=list(projection)
            )
        elif self.querytype == "xquery":
            classads = schedd.xquery(
                requirements=requirements, projection=list(projection)
            )
        else:
            assert False, "Invalid querytype %r" % self.querytype

        data = []
        ad_dicts = utils.classads_to_dicts(classads)
        for ad in ad_dicts:
            if attribute:
                return ad.get(attribute.lower(), None)
            job_data = dict()
            job_data["classad"] = ad
            job_data["jobid"] = "%s.%s" % (ad["clusterid"], ad["procid"])
            data.append(job_data)
        if attribute:
            abort(404, message="No matching ads")
        return data

    def get(self, clusterid=None, procid=None, attribute=None):
        parser = reqparse.RequestParser(trim=True)
        parser.add_argument("projection", default="")
        parser.add_argument("constraint", default="")
        args = parser.parse_args()
        return self.query(
            clusterid,
            procid,
            projection=args.projection,
            constraint=args.constraint,
            attribute=attribute,
        )


class V1JobsResource(JobsBaseResource):
    """Endpoints for accessing information about jobs in the queue

    This implements the following endpoint:

        GET /v1/jobs{/clusterid}{/procid}{/attribute}{?projection,constraint}

        If `clusterid`, `procid`, and `attribute` are specified, then it
        returns the value of that attribute.  Otherwise it returns an array
        of one or more objects of the form:

            {
              "jobid": "123.45",
              "classad": { (json-encoded classad object) }
            }

        If `clusterid` and `procid` are specified, then the array will contain
        a single job.  If only `clusterid` is specified, then the array will
        contain all jobs within that cluster.  If none of these are specified,
        the array will contain all jobs in the queue.

        `projection` is one or more comma-separated attributes; if specified,
        only those attributes, plus `clusterid` and `procid` will be in the
        `classad` object of each job.  `projection` is ignored if `attribute`
        is specified.

        `constraint` is a classad expression restricting which jobs to include
        in the result.  The constraint is always applied, even if `clusterid`
        and `procid` are specified.

    """

    querytype = "xquery"


class V1HistoryResource(JobsBaseResource):
    """Endpoints for accessing historical job information

    This implements the following endpoint:

        GET /v1/history{/clusterid}{/procid}{/attribute}{?projection,constraint}

        If `clusterid`, `procid`, and `attribute` are specified, then it
        returns the value of that attribute.  Otherwise it returns an array
        of one or more objects of the form:

            {
              "jobid": "123.45",
              "classad": { (classad object) }
            }

        If `clusterid` and `procid` are specified, then the array will contain
        a single job.  If only `clusterid` is specified, then the array will
        contain all jobs within that cluster.  If none of these are specified,
        the array will contain all jobs in the history.

        `projection` is one or more comma-separated attributes; if specified,
        only those attributes, plus `clusterid` and `procid` will be in the
        `classad` object of each job.  `projection` is ignored if `attribute`
        is specified.

        `constraint` is a classad expression restricting which jobs to include
        in the result.  The constraint is always applied, even if `clusterid`
        and `procid` are specified.

    """

    querytype = "history"


class V1StatusResource(Resource):
    """Endpoints for accessing condor_status information

    This implements the following endpoint:

        GET /v1/status{?projection,constraint,query}

        This returns an array of objects of the following form:

            {
              "name": "<name classad attribute>",
              "classad": { <classad object> }
            }

        `name` is a specific host or slot to query.  If not specified, all
        matching ads are returned.

        `query` is the type of ad to query; see the "Query options" in the
        condor_status(1) manpage.  "startd" is the default.

        `projection` is one or more comma-separated attributes; if specified,
        only those attributes, plus `name` and `procid` will be in the
        `classad` object of each job.

        `constraint` is a classad expression restricting which ads to include
        in the result.

        Returns 404 if no matching ads are found.  This includes zero ads
        matching the constraint.

    """

    AD_TYPES_MAP = {
        "accounting": AdTypes.Accounting,
        "any": AdTypes.Any,
        "collector": AdTypes.Collector,
        "credd": AdTypes.Credd,
        "defrag": AdTypes.Defrag,
        "generic": AdTypes.Generic,
        "grid": AdTypes.Grid,
        "had": AdTypes.HAD,
        "license": AdTypes.License,
        "master": AdTypes.Master,
        "negotiator": AdTypes.Negotiator,
        "schedd": AdTypes.Schedd,
        "startd": AdTypes.Startd,
        "submitter": AdTypes.Submitter,
        "submitters": AdTypes.Submitter,  # Original API & command-line tools used "submitters"
    }

    def get(self, name=None):
        """GET handler"""
        parser = reqparse.RequestParser(trim=True)
        parser.add_argument("projection", default="")
        parser.add_argument("constraint", default="")
        parser.add_argument(
            "query", choices=list(self.AD_TYPES_MAP.keys()), default="any"
        )
        args = parser.parse_args()

        collector = Collector()
        ad_type = self.AD_TYPES_MAP[args.query]
        projection = []

        if args.projection:
            if not validate_projection(args.projection):
                abort(
                    400,
                    message="Invalid projection: must be a comma-separated list of classad attributes",
                )
            projection = ",".split(args.projection)

        constraint = args.constraint
        if name:
            constraint = '(name == "%s")' % name
            if args.constraint:
                constraint += " && (%s)" % args.constraint

        classads = []  # type: List[classad.ClassAd]
        try:
            classads = collector.query(
                ad_type, constraint=constraint, projection=projection
            )
        except RuntimeError as err:
            abort(400, message=str(err))  # LAZY

        data = [
            {"name": ad["name"], "classad": ad}
            for ad in utils.classads_to_dicts(classads)
        ]

        return data


class V1ConfigResource(Resource):
    """Endpoints for accessing condor config

    This implements the following endpoint:

        GET /v1/config{/attribute}{?daemon}

        If `attribute` is specified, returns the value of the specific
        attribute in the condor config.  If not specified, returns an object
        of the form:

            {
              "attribute1": "value1",
              "attribute2": "value2",
              ...
            }

        If `daemon` is specified, query the given running daemon; otherwise,
        query the static config files.

        Returns 404 if `attribute` is specified but the attribute is undefined.

    """

    DAEMON_TYPES_MAP = {
        "collector": DaemonTypes.Collector,
        "master": DaemonTypes.Master,
        "negotiator": DaemonTypes.Negotiator,
        "schedd": DaemonTypes.Schedd,
        "startd": DaemonTypes.Startd,
    }

    def get(self, attribute=None):
        """GET handler"""
        parser = reqparse.RequestParser(trim=True)
        parser.add_argument("daemon", choices=list(self.DAEMON_TYPES_MAP.keys()))
        args = parser.parse_args()

        if args.daemon:
            daemon_ad = Collector().locate(self.DAEMON_TYPES_MAP[args.daemon])
            param = RemoteParam(daemon_ad)
        else:
            htcondor.reload_config()
            param = htcondor.param

        param_lower = utils.deep_lcasekeys(param)

        if attribute:
            if not validate_attribute(attribute):
                abort(400, message="Invalid attribute")

        if attribute:
            try:
                return param_lower[attribute.lower()]
            except KeyError as err:
                abort(404, message=str(err))

        return param_lower


class RootResource(Resource):
    def get(self):
        return {}


api.add_resource(RootResource, "/")


api.add_resource(
    V1JobsResource,
    "/v1/jobs",
    "/v1/jobs/<int:clusterid>",
    "/v1/jobs/<int:clusterid>/<int:procid>",
    "/v1/jobs/<int:clusterid>/<int:procid>/<attribute>",
)
api.add_resource(
    V1HistoryResource,
    "/v1/history",
    "/v1/history/<int:clusterid>",
    "/v1/history/<int:clusterid>/<int:procid>",
    "/v1/history/<int:clusterid>/<int:procid>/<attribute>",
)
api.add_resource(V1StatusResource, "/v1/status", "/v1/status/<name>")
api.add_resource(V1ConfigResource, "/v1/config", "/v1/config/<attribute>")

PUBLIC_ENDPOINTS = [
    "V1JobsResource",
    "V1ConfigResource",
    "V1HistoryResource",
    "V1StatusResource",
]
